### PR TITLE
[reland] [inductor] fix a max-autotune rng state related bug

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111381

reland https://github.com/pytorch/pytorch/pull/109828

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by desertfire — COMMENTED** [review]
What was changed from the last try?

**Comment by shunting314** [comment]
> What was changed from the last try?

The only change is to make the code work even if numpy is not installed as @lezcano  mentioned in the old PR.

Nothing is changed specific to the failed check encountered in the previous PR. Here is my hypothesis of why the check fail in the old PR: https://github.com/pytorch/pytorch/pull/109828#issuecomment-1732155167 . I don't think that's caused by my code change. The reland passes all tests can somehow back that guess.

File: torch/_dynamo/testing.py [hunk 1]
```diff
@@ -3,13 +3,19 @@
...
 import sys
 import types
 import unittest
 from typing import Sequence, Union
 from unittest.mock import patch
 
+try:
+    import numpy as np
+except ModuleNotFoundError:
+    np = None
```
Comment (lezcano) [inline torch/_dynamo/testing.py:17]: yep

**Review by lezcano — APPROVED** [review]
The NumPy changes LGTM.

**Comment by shunting314** [comment]
Hmm, I see the ROCM  failure again. Looks like the check is only performed at push time. @huydhn  any suggestion how I can get a dev environment to debug it?

**Comment by pruthvistony** [comment]
@jataylo ,
FYI
2023-10-17T19:46:26.0389524Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_inductor/kernel/mm.py", line 259, in tuned_addmm
2023-10-17T19:46:26.0389696Z     return autotune_select_algorithm(
2023-10-17T19:46:26.0390612Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_inductor/select_algorithm.py", line 980, in autotune_select_algorithm
2023-10-17T19:46:26.0390842Z     return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
2023-10-17T19:46:26.0391592Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_inductor/select_algorithm.py", line 712, in __call__
2023-10-17T19:46:26.0391743Z     raise RuntimeError(
2023-10-17T19:46:26.0392151Z torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2023-10-17T19:46:26.0393090Z RuntimeError: No choices to select, please consider adding ATEN into max_autotune_gemm_backends config (defined in torch/_inductor/config.py) to allow at least one choice. 
2023-10-17T19:46:26.0393098Z 
2023-10-17T19:46:26.0393440Z Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
2023-10-17T19:46:26.0393446Z 
2023-10-17T19:46:26.0393452Z 
2023-10-17T19:46:26.0393781Z You can suppress this exception and fall back to eager by setting:
2023-10-17T19:46:26.0393933Z     import torch._dynamo
2023-10-17T19:46:26.0394148Z     torch._dynamo.config.suppress_errors = True
2023-10-17T19:46:26.0394154Z 
2023-10-17T19:46:26.0394159Z 
2023-10-17T19:46:26.0394480Z To execute this test, run the following from the base repo dir:
2023-10-17T19:46:26.0394897Z     PYTORCH_TEST_WITH_ROCM=1 python test_max_autotune.py -k test_cat_addmm
2023-10-17T19:46:26.0394904Z 
2023-10-17T19:46:26.0395275Z This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

**Comment by jataylo** [comment]
Initial thoughts... this looks like expected behaviour for the gfx906 runners.
```
Warning: 17 17:20:41,274] [0/0] torch._inductor.utils: [WARNING] not enough SMs to use max_autotune_gemm mode
```

The max_autotune mode cannot be enabled on this hardware due to the current `is_big_gpu` logic defined in https://github.com/pytorch/pytorch/blob/main/torch/_inductor/utils.py meaning that triton will not be used for gemms.

These tests explicitly list triton as the backend for the max_autotune mode and seems to fail as it will not accept ATen lowered solutions. I am expecting these tests to pass on MI210 runners @jithunnair-amd. There was actually previously a issue disabling these tests for ROCm which seems to have been ignored for the latest runs
https://github.com/pytorch/pytorch/issues/106026 & https://github.com/pytorch/pytorch/issues/106451

I have put out this PR to recommend conditionalising the text_max_autotune UT suite on the same condition as test_select_algorithm to resolve this https://github.com/pytorch/pytorch/pull/111467

**Pull request closed without merging**